### PR TITLE
Add kernel independence test (conditional and unconditional)

### PR DESCRIPTION
**Pull request by bloebp**
Co-authored-by: mgong2
Co-authored-by: kunwuz
Co-authored-by: zhi-yi-huang
Co-authored-by: JarodYv
Co-authored-by: MarkDana
Co-authored-by: cogito233

Changes proposed in this pull request:
Adding HSIC kernel independence test. This is copied and modified from https://github.com/py-why/causal-learn/tree/main/causallearn/utils/KCI.

This is a first proposal how this could look like and how we can add co-authorship to the original authors.

## Before submitting

- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/pywhy-stats/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/pywhy-stats/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.
- [ ] I have added a changelog entry succintly describing the change in this PR in the [whats_new](https://github.com/py-why/pywhy-stats/blob/main/docs/whats_new/) relevant version document.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

File: pywhy_stats/custom_kernels.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+def _get_gaussian_empirical_kci(X: np.ndarray) -> float:
+    n = shape(X)[0]
+    if n < 200:
+        width = 1.2
+    elif n < 1200:
+        width = 0.7
+    else:
+        width = 0.4
+    theta = 1.0 / (width ** 2)
+    return theta / X.shape[1]
```
Comment (adam2392) [inline pywhy_stats/custom_kernels.py:79]: Do we have the exps to back up these numbers? If so, it would be great to also port this into a `benchmarks/` folder and add a link to relevant paper/webpage/etc.

If not, then I would be cautious cuz idk where these came from, so some user might ask us why it defaults to these values and we would have no idea.

File: pywhy_stats/custom_kernels.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+def _get_gaussian_empirical_hsic(X: ndarray):
+    n = shape(X)[0]
+    if n < 200:
+        width = 0.8
+    elif n < 1200:
+        width = 0.5
+    else:
+        width = 0.3
+    theta = 1.0 / (width ** 2)
+    return theta * X.shape[1]
```
Comment (adam2392) [inline pywhy_stats/custom_kernels.py:91]: Same comment as above.

File: pywhy_stats/kci.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+from pywhy_stats.custom_kernels import rbf_kernel
+from pywhy_stats.pvalue_result import PValueResult
+
+
+def ind(X: np.ndarray,
+        Y: np.ndarray,
+        kernel_X: Callable[[np.ndarray], np.ndarray] = rbf_kernel,
+        kernel_Y: Callable[[np.ndarray], np.ndarray] = rbf_kernel,
+        approx: bool = True,
+        null_ss: int = 1000) -> PValueResult:
```
Comment (adam2392) [inline pywhy_stats/kci.py:18]: Should we also expose any keyword arguments to the kernel function? e.g. `**kernel_X_kwargs` and ``*kernel_Y_kwargs``, or `kwidth_X`, etc.

Not sure, so just asking here...

File: pywhy_stats/kci.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    """
+    Python implementation of Kernel-based Conditional Independence (KCI) test and the unconditional version.
+    The original Matlab implementation can be found in http://people.tuebingen.mpg.de/kzhang/KCI-test.zip
+
+    References
+    ----------
+    [1] K. Zhang, J. Peters, D. Janzing, and B. Schölkopf,
+    "A kernel-based conditional independence test and application in causal discovery," In UAI 2011.
+    [2] A. Gretton, K. Fukumizu, C.-H. Teo, L. Song, B. Schölkopf, and A. Smola, "A kernel
+    Statistical test of independence." In NIPS 21, 2007.
```
Comment (adam2392) [inline pywhy_stats/kci.py:28]: You can use the Sphinx bibtex extension, which replaces these with `.. footbibliography::` and then you can reference them in any function as `:footcite:`<name>``. 

See: https://github.com/py-why/pywhy-stats/blob/main/doc/references.bib
and FisherZTest, which shows the pattern.

This just makes the citations of stuff more consolidated especially for papers that are referenced across multiple functions/classes.

File: pywhy_stats/kci.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+from numpy.random import permutation
+from scipy import stats
+
+from pywhy_stats.custom_kernels import rbf_kernel
+from pywhy_stats.pvalue_result import PValueResult
+
+
+def ind(X: np.ndarray,
+        Y: np.ndarray,
+        kernel_X: Callable[[np.ndarray], np.ndarray] = rbf_kernel,
```
Comment (adam2392) [inline pywhy_stats/kci.py:15]: Should we use `numpy.typing.ArrayLike` instead, which is their supported mypy typing extension? 

Ref: https://numpy.org/devdocs/reference/typing.html

**Review by adam2392 — COMMENTED** [review]
Some high level comments first.

File: pywhy_stats/kci.py [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+        kernel_Y: Callable[[np.ndarray], np.ndarray] = rbf_kernel,
```
Comment (adam2392) [inline pywhy_stats/kci.py:16]: We could support passing in a Callable, but also a `str`, which is of any possibility in sklearn's pairwise kernel metrics. This ensures that we have a wide-range of kernels for free. 

E.g. see: https://hyppo.neurodata.io/api/generated/hyppo.independence.hsic

File: pywhy_stats/custom_kernels.py (hunk 1 above)
Comment (bloebp) [reply pywhy_stats/custom_kernels.py:79]: These are just copied over form the original implementation https://github.com/py-why/causal-learn/blob/main/causallearn/utils/KCI/GaussianKernel.py#L39. As far as I know, they were already used in the origin implementation from R.

File: pywhy_stats/custom_kernels.py (hunk 2 above)
Comment (bloebp) [reply pywhy_stats/custom_kernels.py:91]: See previous comment.

File: pywhy_stats/kci.py (hunk 3 above)
Comment (bloebp) [reply pywhy_stats/kci.py:18]: No, the idea is that, if a user wants to customize the kernel calls, they can just provide it via partial. E.g., 
```
kci.ind(X, Y, kernel_X = partial(rbf_kernel, width=1.4, my_other_rbf_parameter=...)
```
This way, we avoid keeping track of all kind of parameters for all kind of kernels.

File: pywhy_stats/custom_kernels.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+        if width == "median":
+            width = _get_gaussian_median_width(X)
+        elif width == "empirical_hsic":
+            width = _get_gaussian_empirical_hsic(X)
+        elif width == "empirical_kci":
+            width = _get_gaussian_empirical_kci(X)
+        else:
+            raise ValueError("Unknown width approximation method!")
+
+    return exp(-0.5 * squareform(pdist(X, 'sqeuclidean')) * width)
```
[the remark this answers is not in the document]
Comment (bloebp) [reply pywhy_stats/custom_kernels.py:38]: Was thinking of using the scikit learn version, but thought it is so trivial and not really much faster that it is not worth it. However, now thinking about it, we can probably even avoid this function completely by pre-configure the kernel via partial (see the other comment) and set the default width method there.

File: pywhy_stats/kci.py (hunk 5 above)
Comment (bloebp) [reply pywhy_stats/kci.py:15]: The issue I see with `ArrayLike` is that the IDE complains that it doesn't have properties such as `shape`, which means, it is not the same as an `ndarray`?

File: pywhy_stats/kci.py (hunk 6 above)
Comment (bloebp) [reply pywhy_stats/kci.py:16]: If a user want's to use a scikit kernel, simply pass it by:
```
from sklearn.metrics.pairwise import polynomial_kernel
ind(X, Y, kernel_X = polynomial_kernel)
```
thats all they need to do. And we avoid having a long list of possible strings.

Comment (bloebp) [reply pywhy_stats/kci.py:16]: Another advantage is, the kernels can also be pre-configure via partials, which will not be possible if we just allow a string to select it.

Comment (adam2392) [reply pywhy_stats/kci.py:16]: > If a user want's to use a scikit kernel, simply pass it by:
> 
> ```
> from sklearn.metrics.pairwise import polynomial_kernel
> ind(X, Y, kernel_X = polynomial_kernel)
> ```
> 
> thats all they need to do. And we avoid having a long list of possible strings.

I like that. This is a good example to include in the `Examples` section in the docstring.

File: pywhy_stats/kci.py (hunk 5 above)
Comment (adam2392) [reply pywhy_stats/kci.py:15]: That's weird. Shapes hasn't been implemented yet https://github.com/numpy/numpy/issues/16544, but maybe there's some issue w/ the IDE? 

But I think ArrayLike is more general because it will supposedly support any "array-API" (in the future). But yeah not super important until shapes implemented imo.

File: pywhy_stats/kci.py (hunk 3 above)
Comment (adam2392) [reply pywhy_stats/kci.py:18]: Okay. I think I'm on board. It would be great to also have this pattern as a MWE in the `Examples` docstring. Then users can easily copy-paste the pattern.

File: pywhy_stats/custom_kernels.py (hunk 1 above)
Comment (adam2392) [reply pywhy_stats/custom_kernels.py:79]: Tagging @kunwuz.

Any idea where these values came from? Anyway, we can inform ppl?

Comment (kunwuz) [reply pywhy_stats/custom_kernels.py:79]: They are from the second paragraph of section 3.5 in the [KCI paper ](https://arxiv.org/ftp/arxiv/papers/1202/1202.3775.pdf). I agree that including a reference could be helpful for these empirical values.

**Comment by bloebp** [comment]
Replacing this with: https://github.com/py-why/pywhy-stats/pull/14
